graph_weather/data/sampler.py

```diff
@@ -0,0 +1,188 @@
+import logging
+import os
+import random
+
+import torch
+import xarray as xr
+
+
+class HierarchicalSampler:
+    def __init__(self, config: dict, data_path: str, loader=None):
+        """
+        Generic Hierarchical Data Sampler.
+
+        Args:
+            config (dict): Configuration parameters, e.g.:
+                - input_fields: List[str], fields to load.
+                - time_segment_samples: int, number of time segments to sample.
+                - time_slices_per_segment: int, number of time slices per segment.
+                - time_steps: int, number of consecutive time steps to extract.
+                - neighborhoods_per_slice: tuple(int, int), range for the number of spatial patches.
+                - neighborhood_size: tuple(int, int), patch size in (latitude, longitude).
+                - patch_size: int, size for tokenization patches.
+                - mask_ratio: float, fraction of tokens to mask.
+                - years: tuple(int, int), (start_year, end_year); used for default time segments.
+                - file_pattern: str, pattern to locate files (default: "{field}_{year}_{month:02d}.nc").
+            data_path (str): Base directory containing the data files.
+            loader (callable, optional): Custom function for loading a dataset. Should accept a time_segment and field,
+                and return an xarray.Dataset (or None if not found). If not provided, a default loader is used.
+        """
+        self.config = config
+        self.data_path = data_path
+        self.loader = loader if loader is not None else self._default_loader
+
+        # Create a list of available time segments (e.g., year-month pairs)
+        self.time_segments = self._get_available_time_segments()
+        self.logger = logging.getLogger("HierarchicalSampler")
+        self.logger.setLevel(logging.INFO)
+
+    def _get_available_time_segments(self):
```

Comment:
This isn't actually checking available times in the dataset, just generating a list of years and months.